### DISCUSSION TITLE
How to initialize a vector of stateful(old) materials?

### DISCUSSION THREAD

**Discussion by xueyang94** [idaholab/moose#19209] [Q&A Modules: General] [answered]
Hello,

I am writing a material class where I want to initialize a vector of stateful materials. What I did is as follows:

- In the validParams list, add `params.addRequiredParam<std::vector<MaterialPropertyName>>("some_name", "a description");`. 
- In the constructor, add `_some_old(getMaterialPropertyOld<std::vector<Real>>("some_name"))`. 
- In the header file, declare `const MaterialProperty<std::vector<Real>> & _some_old`. 

Now in computeQpProperties, I can use `_some_old[_qp][i]` to directly access the old values without any compiler error. However, I received an error when running the input file: `The following error occurred in the object "MOOSE Problem", of type "FEProblem". One or more Material Properties were not supplied on block 0: some_name`, even though I have supplied `some_name = 'name1, name2, ...'` in the material class block. Could anyone please let me know there this error comes from, and how to fix it? Thanks in advance.

**Comment by GiudGiud** [comment, marked as answer]
Hello

The issue is that you adding in the Params a vector of material properties. 
But in the constructor you are trying to retrieve a material property that is vector-valued. 

Based on your other post, I think you are trying to specify the vector as "mat1 mat2 ...". This cant be done. You have to declare a material property of type std::vector<Real> first, then fill it in another material with the mat1, mat2 etc values. Then you can use it here. 

Also, prefer to use RealVectorValue over std::vector < Real > for the data type.

**Reply by xueyang94** [reply]
Thanks for the reply. I actually tried to initialize a vector of old materials the same way as I did in the other post for non-old materials, but the syntax is wrong. For non-old materials, we can do `_some_name(getParam<std::vector<MaterialPropertyName>>("some_name"))` in the initializer, and also have a `_some_prop` to fill the values in like you said. However, with old material, I tried `_some_old(getMaterialPropertyOld<std::vector<MaterialPropertyName>>("some_name"))`, but the compiler error is that there is no matching constructor. How should the old material vector syntax be? Appreciate it.

**Reply by GiudGiud** [reply]
The non-old syntax would be:
~~`_some(getMaterialProperty<std::vector<MaterialPropertyName>>("some_name"))`~~
`_some(getMaterialProperty< std::vector < Real > >("some_name"))`

this
`_some_name(getParam<std::vector<MaterialPropertyName>>("some_name"))`
is completely different

**Reply by xueyang94** [reply]
I am using the getParam for non-old material and it works fine. It is the same as I mentioned in my previous post. The getMaterialProperty won't compile. But either way, what would you suggest to use for old material syntax? `_some_old(getMaterialPropertyOld<std::vector<MaterialPropertyName>>("some_name"))` does not work...

**Reply by GiudGiud** [reply]
it works to populate a vector of string. It doesnt get you the material property.

If you have a vector material property, this is the syntax
~~`_some_old(getMaterialPropertyOld<RealVectorValue>("some_name"))`~~
`_some_old(getMaterialPropertyOld< std::vector < Real > >("some_name"))`

**Reply by dschwen** [reply]
`RealVectorValue` stores 3-vectors. If Xueyang wants to store a vector of 10 or 20 values (which we have plenty of precedent for), then a property of type `std::vector<Real>` would be appropriate. But there's a lot of confusion in this thread (for example a material property of type `std::vector<MaterialPropertyName>` ... what?!)

**Reply by GiudGiud** [reply]
oh yeah my bad about RealVectorValue and not noticing the syntax had gotten bad since the initial post. The rest is still valid.

**Reply by dschwen** [reply]
I'm assuming you mean `some_name` to be a single name of a material property that contains a vector of values (at least thats how you use the parameter. That means it should be

```
params.addRequiredParam<MaterialPropertyName>("some_name", "a description");
```

**Reply by xueyang94** [reply]
@dschwen Hi Daniel, `some_name` is a single name of material property that contains a vector of material property names, not real values. To be specific, it stores the vector of phase concentration names in the KKS local solve. The names of the phase concentrations are user-supplied. For example, in the input file material block, we would have `ci_names = 'c1, c2, ...'`. Of course, c1 and c2 are computed in my material class. 

I have figured out how to declare and initialize `ci_names` as a non-old vector of ci (I put it in [this post](https://github.com/idaholab/moose/discussions/19152)), but I am stuck on the old material vector. I want to initialize `_ci_old` as a vector of old ci in the material source code, but I am not sure how to use the getMaterialPropertyOld in this case.

**Reply by GiudGiud** [reply]
Has this been solved?
getMaterialPropertyOld should be called with the name of the material property. If it s been made a vector then the name of that one. 

If it's some component like c1, c2 etc then they should be retrieved one by one or made into a vector material property first
